packages/EasyCodingStandard/packages/SniffRunner/src/Parser/FileToTokensParser.php

```diff
@@ -39,7 +40,14 @@ public function createTokenizerFromFileInfo(SmartFileInfo $smartFileInfo): PHP
     {
         $fileContent = $this->cachedFileLoader->getFileContent($smartFileInfo);
 
-        return new PHP($fileContent, $this->getLegacyConfig(), PHP_EOL);
+        return new PHP($fileContent, $this->getLegacyConfig(), Common::detectLineEndings($fileContent));
+    }
+
+    public function detectLineEndingsFromFileInfo(SmartFileInfo $smartFileInfo): string
+    {
+        $fileContent = $this->cachedFileLoader->getFileContent($smartFileInfo);
+
+        return Common::detectLineEndings($fileContent);
```

Comment:
Is `detectLineEndings` cached? Here it's called twice for the same file.

Reply:
Is this a trick question? :) I guess you have to see that from those 11 lines by yourself.

`detectLineEndings` itself is not cached. So it does twice the same search. Tbh it doesn't make sense to me to cache it (=saving it to some centralized place from which we can load it later). It would be sensible to detect it only once and then pass it through method arguments.

This method (`createFromFileInfo`) creates the same tokenizer twice (by calling `parseFromFileInfo` and `createTokenizerFromFileInfo`), which does the same thing - loads file content from cache. So even before my change there was loading from cache twice. Now it loads it three times. 

So if wasn't a problem before, I can't see it as a problem now. But of course it's a room for some optimalization. This just seemed to be the best way without changing method signatures and possibly making some BC breaks.

Reply:
It's normal question, performance is important :) 

If `detectLineEndings` inside is not cached, I'd like to see performance comparison on ~100 files with psr2 set - before and after this PR.

If it's small number, we can merge this.

Reply:
It just seems to me impossible not to see from those 11 lines that there is no caching. No offense. :) I do understand those performance concerns (otherwise I wouldn't notice that the tokenizer is also initialized twice).

I will try to prepare it tomorrow. I would appreciate tips or examples so it's more to your liking (only if you have something, otherwise I'll manage).

Reply:
I don't see inside `detectLineEndings` in this PR.

Reply:
I see now, thank you for clarification. It didn't occur to me to consider that part as a potential "performance hog". I personally think loading the content from file cache is more demanding than that. `detectLineEndings` just seaches for `\r\n` using `preg_match` in file content (string).

---

I used whole `ikeblaster/Symplify` for benchmark (working dir is `Symplify/packages/EasyCodingStandard/`). ECS progress bar reports 788 files. Used `clear-cache` option (cached "clean" files are not affected by this change). SATA SSD, CPU i7-6700.

`benchmark.sh:`
```
bin/ecs -V

time bin/ecs check ../../ --config ./config/psr2.yml --no-ansi --no-interaction --no-progress-bar --clear-cache > /dev/null
```

Before change (checkout of tag v5.4.1):
```
$ ./benchmark.sh
EasyCodingStandard 5.4.1.x-dev@ab6785a

real    0m25.070s
user    0m0.060s
sys     0m0.184s

$ ./benchmark.sh
EasyCodingStandard 5.4.1.x-dev@ab6785a

real    0m25.035s
user    0m0.015s
sys     0m0.184s

$ ./benchmark.sh
EasyCodingStandard 5.4.1.x-dev@ab6785a

real    0m24.705s
user    0m0.091s
sys     0m0.122s
```

Then checkout of master and `composer install --no-dev` (to regenerate versions file). Results of the same script:
```
$ ./benchmark.sh
EasyCodingStandard 5.5.x-dev@b506cd6

real    0m26.889s
user    0m0.031s
sys     0m0.060s

$ ./benchmark.sh
EasyCodingStandard 5.5.x-dev@b506cd6

real    0m25.567s
user    0m0.092s
sys     0m0.139s

$ ./benchmark.sh
EasyCodingStandard 5.5.x-dev@b506cd6

real    0m25.559s
user    0m0.077s
sys     0m0.090s
```

So in general it's 1 - 1.5 seconds slower on 788 files (this corresponds to the first check of a cloned project, when you have empty cache).